File: src/tresjolie/geodesy.clj
(ns tresjolie.geodesy)

; Playing around with geodesy,
; based on http://www.movable-type.co.uk/scripts/latlong.html,
; Recipe 1.18 "Performing Trigonometry" in "Clojure Cookbook",
; and Wikipedia (http://en.wikipedia.org/wiki/Earth_radius#Geocentric_radius).

; Earth's approximate radius at latitude 61.50
(def earth-radius 6371.009)

(def earth-equatorial-radius 6378.1370)
(def earth-polar-radius 6356.7523)
    
; Converts all the values in point from degrees to radians.
; The mapv function (in clojure.core since Clojure 1.4) applies
; the function to all the members of a collection and returns
; a vector. The #() is a function literal with one argument.
; Thus, the return value of this function is a vector with the
; results of applying Math.toRadians to all members of `point`.
; You would call this function like so:
; (degrees->radians [61.5, 23.75, 73.2, 2.43])
(defn degrees->radians [point]
    (mapv #(Math/toRadians %) point))

(defn square [x] (* x x))

(defn term [a b] (+ (square a) (square b)))

(defn term1-a [lat] (* (square earth-equatorial-radius) (Math/cos lat)))
(defn term1-b [lat] (* (square earth-polar-radius) (Math/sin lat)))
(defn term1 [lat] (+ (square (term1-a lat)) (square (term1-b lat))))

(defn term2-a [lat] (* (earth-equatorial-radius (Math/cos lat))))
(defn term2-b [lat] (* (earth-polar-radius (Math/cos lat))))
(defn term2 [lat] (+ (square term2-a) (square term2-b)))

(defn earth-geocentric-radius 
    "Calculate the geocentric radius of Earth at the given latitude,
     in degrees."
    [lat] (Math/sqrt (/ 
        (+ (square (* (square earth-equatorial-radius) (Math/cos lat))) (square (* (square earth-polar-radius) (Math/sin lat))))
        (+ (square (* (square earth-equatorial-radius) (Math/cos lat))) (square (* (square earth-polar-radius) (Math/sin lat))))
        )))

(defn distance-between
    "Calculate the distance in km between two points on Earth. Each
     point is a pair of degrees latitude and longitude, in that order."
    ([p1 p2] (distance-between p1 p2 earth-radius))
    ([p1 p2 radius]
        (let [[lat1 long1] (degrees->radians p1) 
              [lat2 long2] (degrees->radians p2)]
            (* radius
                (Math/acos (+ (* (Math/sin lat1) (Math/sin lat2))
                              (* (Math/cos lat1)
                                 (Math/cos lat2)
                                 (Math/cos (- long1 long2)))))))))

; "Clojure Cookbook" by Luke VanderHart and Ryan Neufeld (O’Reilly).
; Copyright 2014 Cognitect, Inc., 978-1-449-36617-9.”"Clojure Cookbook"

(defn within-distance
    "Predicate for finding out if the given point is within a distance of another.
     The distance is given as a double value in kilometers."
    ([dist p1 p2] (<= (distance-between p1 p2) dist)))

File: python/tresjolie/geodesy.py
import math

EARTH_EQUATORIAL_RADIUS = 6378.1370
EARTH_POLAR_RADIUS = 6356.7523

#
# Utility functions
#
 
def geocentric_radius(latitude_degrees):
    """ 
    Computes the geocentric radius of the Earth for the given latitude.
    
    See http://en.wikipedia.org/wiki/Earth_radius#Geocentric_radius.
    """    
    latitude = math.radians(latitude_degrees)
    a = EARTH_EQUATORIAL_RADIUS
    b = EARTH_POLAR_RADIUS

    t1a = a**2 * math.cos(latitude)
    t1b = b**2 * math.sin(latitude)
    t1 = t1a**2 + t1b**2
    
    t2a = a * math.cos(latitude)
    t2b = b * math.sin(latitude)
    t2 = t2a**2 + t2b**2
    
    r = math.sqrt(t1 / t2)
    return r


def haversine(lat1_deg, lon1_deg, lat2_deg, lon2_deg, earth_radius=EARTH_EQUATORIAL_RADIUS):
    """
    Computes the distance between two geographical points in kilometers.
    
    The coordinate values are degrees.
    Uses the haversine formula (see http://en.wikipedia.org/wiki/Haversine_formula).
    This implementation is derived from the JavaScript code 
    by Andrew Hedges at http://andrew.hedges.name/experiments/haversine/.
    The Earth radius defaults to the Earth's equatorial radius.
    For added accuracy, you may want pass a radius more suitable for the
    latitudes of the points you are processing.
    """    
    lat1 = math.radians(lat1_deg)
    lon1 = math.radians(lon1_deg)
    lat2 = math.radians(lat2_deg)
    lon2 = math.radians(lon2_deg)
    
    dlon = lon2 - lon1 
    dlat = lat2 - lat1 
    a = (math.sin(dlat/2))**2 + math.cos(lat1) * math.cos(lat2) * (math.sin(dlon/2))**2 
    c = 2 * math.atan2( math.sqrt(a), math.sqrt(1-a) ) 
    d = earth_radius * c

    return d
